### PR TITLE
Revise CachePolicy based on what we've learned.

### PR DESCRIPTION
**Pull request by toddmgreer**
Commit Message:
Revise CachePolicy interface based on experience using a forked version.
Additional Description:
- CachePolicy can now be stored in FilterState
- CacheEntryUsability:
  - Added ttl for use in cache-status header
  - Added op==, op!= for use in tests
- requestCacheable:
  - Now returns enum, so it can say it's ok to serve from cache but not insert into cahce
  - Renamed to requestCacheability to reflect not returning a bool
- responseCacheable:
  - Now returns enum, so it can say to mark a cache key as uncacheable to enable later optimizations
  - Renamed to responseCacheability to reflect not returning a bool
  - Removed VaryHeader parameter--CacheFilter should only call responseCacheable if the Vary headers are ok
- createCacheKey: renamed to cacheKey, to avoid implying that the key must literally be created here
- Removed *CacheControl parameters--not always needed; CachePolicy implementations can make then if needed
- Removed setCallbacks--CachePolicy implementations that need FilterState should take it as a ctor parameter
Risk Level: None--not yet used
Testing: n/a--it's just an interface
Docs Changes: n/a--not yet used
Release Notes: n/a--not yet used
Platform Specific Features: n/a

### CODE REVIEW

File: source/extensions/filters/http/cache/cache_policy.h
Comment (ravenblackx) [inline source/extensions/filters/http/cache/cache_policy.h:28]: Drop the `Envoy::`, or also include it on the type and on both things in the instance above.

File: source/extensions/filters/http/cache/cache_policy.h
Comment (ravenblackx) [inline source/extensions/filters/http/cache/cache_policy.h:42]: This comment

File: source/extensions/filters/http/cache/cache_policy.h
Comment (ravenblackx) [inline source/extensions/filters/http/cache/cache_policy.h:99]: What's the reason for removing vary_allow_list? Is is that rejections based on that make it so we don't even need to query policy?

**Review by ravenblackx — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/cache/cache_policy.h
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_policy.h]: Fixed.

File: source/extensions/filters/http/cache/cache_policy.h
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_policy.h]: If a response has a `vary` header that isn't in the vary allowlist, there's no need for CacheFilter to ask CachePolicy about it--CacheFilter already has enough information to know not to store it. Thus, there's no need to pass the allowlist to this function.

**Review by ravenblackx — DISMISSED** [review]

File: source/extensions/filters/http/cache/cache_policy.h
Comment (ravenblackx) [inline source/extensions/filters/http/cache/cache_policy.h:44]: Could you elaborate on this comment a bit? I don't understand how you would be able to serve a response from cache if you can't store it.

Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_policy.h:44]: Request 1: "GET /", response gets stored in cache
Request 2: "GET /", but with "cache-control: no-store"

Request 2 can be served from cache, but if it weren't found in cache, we wouldn't[1] store the response.

If all the requests for a particular URL have "cache-control: no-store", then none will ever be stored, but if only some have "no-store", then there may be a cache entry to serve from, even if a particular request can't[1] be stored.

[1] https://httpwg.org/specs/rfc9111.html#cache-request-directive lets caches choose whether to honor request cache-control directives. The products I work on honor "no-store", because that may express a concern that the response could have private information, and that's a concern to always take seriously.

Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_policy.h:44]: Thanks! (Might be worth adding a short note to this effect in the code comment so future people have the same context.)

Also there's a typo requst -> request in the comment above, which I didn't previous mention because CI was mentioning it.

**Review by mkbehr — CHANGES_REQUESTED** [review]
Can you add a summary to the commit description of what's changing and why?

File: source/extensions/filters/http/cache/cache_policy.h
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_policy.h:28]: If a CachePolicy implementation needs FilterState, I expect that it will take a `FilterState&` parameter in its constructor.

File: source/extensions/filters/http/cache/cache_policy.h
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_policy.h:42]: I added them in cache_policy.h, as advised by https://google.github.io/styleguide/cppguide.html#Operator_Overloading: "define them in the same headers, .cc files, and namespaces as the types they operate on."

File: source/extensions/filters/http/cache/cache_policy.h
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_policy.h]: Clarified

**Comment by toddmgreer** [comment]
> Can you add a summary to the commit description of what's changing and why?

I don't know how to do that. A maintainer with merge authority will eventually squash all commits into one, and I don't know how that commit gets its description. I guess we can bring this up with the maintainer once that review gets going.

**Comment by jmarantz** [comment]
Maintainers usually cut and paste from the Deescription to populate the message for the squashed commit.

**Review by ravenblackx — APPROVED** [review]

File: source/extensions/filters/http/cache/cache_policy.h
Comment (mkbehr) [reply source/extensions/filters/http/cache/cache_policy.h:42]: Our other operator<<s for testing purposes live in test/.../common.h, so we should be consistent one way or the other. Looks like they originally got moved over in #12544, and I don't see any claims there that Envoy style requires they live in test libraries, so I agree that it makes sense to put them back here. I'm fine keeping them here if we can quickly send a followup change to move the others.

**Comment by mkbehr** [comment]
> > Can you add a summary to the commit description of what's changing and why?
> 
> I don't know how to do that. A maintainer with merge authority will eventually squash all commits into one, and I don't know how that commit gets its description. I guess we can bring this up with the maintainer once that review gets going.

There's a comment at the top of the pull request that got populated with a template, and IME that's where the description comes from. I think github will let you edit the comment.

This looks good to me so far but I want to make sure I understand everything that's different from our internal changes (e.g. removing the callbacks) before I hit approve, so the summary will help that.

**Review by mkbehr — APPROVED** [review]

**Comment by toddmgreer** [comment]
@jmarantz this is ready for your review

File: source/extensions/filters/http/cache/cache_policy.h
Comment (jmarantz) [inline source/extensions/filters/http/cache/cache_policy.h:48]: I think Raven asked this already but I missed the answer, but  how does a no-store request get served from cache if the response cannot be stored? Maybe you could add that to the comment here?

**Review by jmarantz — COMMENTED** [review]
one minor nit and I think I can merge this, as it's just an extension.

File: source/extensions/filters/http/cache/cache_policy.h
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_policy.h]: Expanded explanation. Is that sufficiently clear?

**Review by jmarantz — DISMISSED** [review]
Thanks!

**Comment by jmarantz** [comment]
Needs format fix

**Comment by toddmgreer** [comment]
Fixed formatting/spelling.

**Review by jmarantz — APPROVED** [review]

**Pull request merged**
